File: realtime-service/pkg/models/operation.go
package models

import "time"

// Operation represents an operational transform operation
type Operation struct {
	Type     string `json:"type"` // "insert", "delete", "retain"
	Position int    `json:"position"`
	Text     string `json:"text,omitempty"`
	UserID   string `json:"userId"`
	Version  int64  `json:"version"`
}

// Document represents a collaborative document
type Document struct {
	ID        string
	ProjectID string
	FilePath  string
	Content   string
	Version   int64
	History   []Operation
	UpdatedAt time.Time
}

File: business-logic-service/app/models/operation.py
"""Operation model for operational transform"""

from datetime import datetime
from typing import Optional
from uuid import UUID, uuid4

from sqlalchemy import String, DateTime, ForeignKey, func, Text, Integer, BigInteger
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Operation(Base):
    """Operation model for document editing history"""

    __tablename__ = "operations"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    document_id: Mapped[UUID] = mapped_column(
        ForeignKey("documents.id", ondelete="CASCADE"), nullable=False, index=True
    )
    user_id: Mapped[UUID] = mapped_column(ForeignKey("users.id"), nullable=False)
    operation_type: Mapped[str] = mapped_column(
        String(50), nullable=False
    )  # insert, delete, retain
    position: Mapped[int] = mapped_column(Integer, nullable=False)
    text: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    version: Mapped[int] = mapped_column(BigInteger, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    def __repr__(self) -> str:
        return f"<Operation(id={self.id}, document_id={self.document_id}, type={self.operation_type}, version={self.version})>"
